File: week4/java/GenerateParentheses_22.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        
        // 1st approach:
        // if (n == 0) {
        //     result.add("");
        // } else {
        //     for (int c = 0; c < n; c++) {
        //         for (String left : generateParenthesis(c)) {
        //             for (String right : generateParenthesis(n - 1 - c)) {
        //                 result.add("(" + left + ")" + right);
        //             }
        //         }
        //     }
        // }
        // return result;
        
        // 2nd approach:
        generate(n, n, "", result);
        return result;
    }
    
    private void generate(int left, int right, String s, List<String> result) {
        if (right < left) {
            return;
        }
        if (left == right && left == 0) {
            result.add(s);
            return;
        }
        if (left > 0) {
            generate(left - 1, right, s + "(", result);
        } 
        if (right > 0) {
            generate(left, right - 1, s + ")", result);
        }
    }
   
}

File: week4/python/GenerateParentheses_22.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        
        result = list()
        
        def generate(count_o: int, count_c: int, paren: str):
            if count_o > count_c:
                return 

            if count_o == count_c and count_o == 0:
                result.append(paren)
                return
            
            if count_o > 0:
                generate(count_o - 1, count_c, paren + "(")
            
            if count_c > 0:
                generate(count_o, count_c - 1, paren + ")")
        
        generate(n, n, "")
        return result
